cpp/daal/include/data_management/data/internal/numeric_table_sycl_csr.h

```diff
@@ -0,0 +1,673 @@
+
+            return;
+        }
+        if (_dataSize)
+        {
+            st |= setArrays(bufferData, bufferColIndices, bufferRowOffsets, indexing);
+        }
+    }
+
+    template <typename T>
+    services::Status getTBlock(size_t idx, size_t nrows, int rwFlag, BlockDescriptor<T> & block)
+    {
+        services::throwIfPossible(services::ErrorMethodNotImplemented);
+        return services::ErrorMethodNotImplemented;
+    }
+
+    template <typename T>
+    services::Status releaseTBlock(BlockDescriptor<T> & block)
+    {
+        if (!(block.getRWFlag() & (int)writeOnly)) block.reset();
+        return services::Status();
+    }
+
+    template <typename T>
+    services::Status getTFeature(size_t feat_idx, size_t idx, size_t nrows, int rwFlag, BlockDescriptor<T> & block)
+    {
+        services::throwIfPossible(services::ErrorMethodNotImplemented);
+        return services::ErrorMethodNotImplemented;
+    }
+
+    template <typename T>
+    services::Status releaseTFeature(BlockDescriptor<T> & block)
+    {
+        services::throwIfPossible(services::ErrorMethodNotImplemented);
+        return services::ErrorMethodNotImplemented;
+    }
+
+    template <typename T>
+    services::Status getSparseTBlock(size_t idx, size_t nrows, int rwFlag, CSRBlockDescriptor<T> & block)
+    {
+        using namespace services::internal::sycl;
+
+        size_t ncols = getNumberOfColumns();
+        size_t nobs  = getNumberOfRows();
+        block.setDetails(ncols, idx, rwFlag);
+
+        if (idx >= nobs)
+        {
+            block.resizeValuesBuffer(0);
+            return services::Status();
+        }
+
+        nrows = (idx + nrows < nobs) ? nrows : nobs - idx;
+
+        services::Status st;
+
+        block.setRowIndicesBuffer(_rowOffsets);
+
+        size_t offset   = 0;
+        size_t datasize = _dataSize;
+        if (idx == 0)
+        {
+            block.setRowIndicesBuffer(_rowOffsets);
+        }
+        else
+        {
+            services::internal::sycl::UniversalBuffer rowOffsetsNew =
+                services::internal::getDefaultContext().allocate(services::internal::sycl::TypeIds::id<size_t>(), (nrows + 1), st);
+            DAAL_CHECK_STATUS_VAR(st);
+            services::SharedPtr<size_t> hostRowOffsetsNew = rowOffsetsNew.get<size_t>().toHost(ReadWriteMode::writeOnly, st);
+            DAAL_CHECK_STATUS_VAR(st);
+            const services::SharedPtr<size_t> hostRowOffsets = _rowOffsets.toHost(ReadWriteMode::readOnly, st);
+            DAAL_CHECK_STATUS_VAR(st);
+
+            size_t * rowOffsetsNewPtr    = hostRowOffsetsNew.get();
+            const size_t * rowOffsetsPtr = hostRowOffsets.get();
+
+            rowOffsetsNewPtr[0] = 1;
+            for (size_t i = 0; i < nrows; ++i)
+            {
+                const size_t nNonZeroValuesInRow = rowOffsetsPtr[idx + i + 1] - rowOffsetsPtr[idx + i];
+                rowOffsetsNewPtr[i + 1]          = rowOffsetsNewPtr[i] + nNonZeroValuesInRow;
+            }
```

Comment:
Do we assume this code is safe in terms of sum overflows?

Reply:
Yes. Because the sum of all nNonZeroValuesInRow be greater than dataSize form original data (by definition CSR)